### PR TITLE
[IMP] account: Logging and tracking of some fields.

### PR DESCRIPTION
**Pull request by aboo-odoo**
Problem
---------
All fields in the tax definition are currently modifiable even after the
tax has been used. This raises some issues in tax reports and make
investigations difficult when things go wrong in tax report.

Objective
---------
Once an account tax has been used:
 - Some fields in the account tax should not be modifiable anymore to
 avoid tax report issues;
 - Other fields should be logged when modified to ease debugging and
 investigations.

Solution
---------
1. Add a logger to the tax form to track the modification of some fields
2. Track whether the tax is being used in transactions or not
3. Make the fields that should not be modified anymore readonly when the
tax is being used.

Task-3450002
---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by tsb-odoo** [comment]
@aboo-odoo We could track the factor_percent field too.
And the is_base_affected field which is only visible in debug mode (advances options tab). Once done, the rest looks like to me and it can go straight to review. .

**Review by rugo-odoo — CHANGES_REQUESTED** [review]
Some comments and you forgot to add a nice/pretty little test :blush:

File: addons/account/models/account_tax.py [hunk 1]
```diff
@@ -114,16 +115,17 @@ class AccountTax(models.Model):
...
     sequence = fields.Integer(required=True, default=1,
         help="The sequence field is used to define order in which the tax lines are applied.")
-    amount = fields.Float(required=True, digits=(16, 4), default=0.0)
+    amount = fields.Float(required=True, digits=(16, 4), default=0.0, tracking=True)
     description = fields.Char(string='Description', translate=True)
     invoice_label = fields.Char(string='Label on Invoices', translate=True)
     price_include = fields.Boolean(string='Included in Price', default=False,
         help="Check this if the price you use on the product and invoices includes this tax.")
-    include_base_amount = fields.Boolean(string='Affect Base of Subsequent Taxes', default=False,
+    include_base_amount = fields.Boolean(string='Affect Base of Subsequent Taxes', default=False, tracking=True,
```
[the remark this answers is not in the document]
Comment (aboo-odoo) [reply addons/account/models/account_tax.py:129]: @rugo-odoo - Actually after discussion with OCO and LAS, the default value of a Boolean is NULL (the ORM does not set anything). If it is important the default value is set, then it must be explicit. This is optimization apparently.

[the remark this answers is not in the document]
Comment (rugo-odoo) [reply addons/account/models/account_tax.py:129]: :exploding_head:

File: addons/account/tests/test_chart_template.py [hunk 2]
```diff
@@ -109,13 +108,6 @@ def _tax_vals(name, amount, tax_tag_id=None, children_tax_xmlids=None, active=Tr
     }
     if children_tax_xmlids:
         tax_vals.update({'children_tax_ids': [Command.set(children_tax_xmlids)]})
-    else:
-        tax_vals.update({'repartition_line_ids': [
-            Command.create({'document_type': 'invoice', 'factor_percent': 100, 'repartition_type': 'base', 'tag_ids': tag_command}),
-            Command.create({'document_type': 'invoice', 'factor_percent': 100, 'repartition_type': 'tax'}),
-            Command.create({'document_type': 'refund', 'factor_percent': 100, 'repartition_type': 'base'}),
-            Command.create({'document_type': 'refund', 'factor_percent': 100, 'repartition_type': 'tax'}),
-        ]})
```
Comment (rugo-odoo) [inline addons/account/tests/test_chart_template.py:118]: Why this change ?

Comment (aboo-odoo) [reply addons/account/tests/test_chart_template.py:118]: The tax restriction after usage creates issues in the chart template
tests: some of those test would re-create the repartition
lines that were created by default after the tax is used which is not
permitted anymore. This removes the Command create that would recreate the lines. When tags are needed (only in 1 test), the test sets it itself.

File: addons/account/models/account_tax.py [hunk 3]
```diff
@@ -224,6 +239,34 @@ def _compute_tax_group_id(self):
                 ('country_id', '=', False),
             ], limit=1)
 
+    def _compute_is_used(self):
+        taxes_in_transactions_ctr = (
+            Counter(dict(self.env['account.move.line']._read_group([], groupby=['tax_ids'], aggregates=['__count']))) +
+            Counter(dict(self.env['account.reconcile.model.line']._read_group([], groupby=['tax_ids'], aggregates=['__count'])))
+        )
+
+        installed_models = self.env['ir.model'].sudo().search_fetch([('model', 'in', list(OTHER_MODELS_USING_TAXES.keys()))], ['model'], limit=len(OTHER_MODELS_USING_TAXES))
```
Comment (qdp-odoo) [inline addons/account/models/account_tax.py:248]: hum... this isn't the odoo way

We'd rather define a function, inherited in the different modules, each of them adding their own model to the list of models we have to check. 
The benefit is that each module then contains all the code and logic it is adding. And it's arguably cleaner :)

**Review by qdp-odoo — CHANGES_REQUESTED** [review]
I think some parts still need some rework. I'm open to discussing any comment.

Tests are nice, though

File: addons/account/models/account_tax.py (hunk 3 above)
Comment (aboo-odoo) [reply addons/account/models/account_tax.py:248]: Indeed, probably forgotten leftovers from testings!

File: addons/account/models/account_tax.py [hunk 4]
```diff
@@ -225,6 +238,84 @@ def _compute_tax_group_id(self):
...
+        self.ensure_one()
+
+        if self.is_used:
+            repartition_line_str_field_id = self.env['ir.model.fields'].search([('name', '=', 'repartition_lines_str')], limit=1).id
+            for tracked_value_id in kwargs['tracking_value_ids']:
+                if tracked_value_id[2]['field_id'] == repartition_line_str_field_id:
+                    kwargs['tracking_value_ids'].remove(tracked_value_id)
+                    self._message_log_repartition_lines(tracked_value_id[2]['old_value_char'], tracked_value_id[2]['new_value_char'])
+
+            return super()._message_log(**kwargs)
```
Comment (qdp-odoo) [inline addons/account/models/account_tax.py:317]: this method needs documentation, and ho, look... there is an OVERRIDE comment that you could use :)

File: addons/account/models/account_tax.py [hunk 5]
```diff
@@ -359,7 +450,14 @@ def _sanitize_vals(self, vals):
 
     @api.model_create_multi
     def create(self, vals_list):
-        return super().create([self._sanitize_vals(vals) for vals in vals_list])
+        context = clean_context(self.env.context)
+        context.update({
+            'mail_create_nosubscribe': True,
+            'mail_auto_subscribe_no_notify': True,
+            'mail_create_nolog': True,
+        })
```
Comment (qdp-odoo) [inline addons/account/models/account_tax.py:458]: need comments: what are those context key for?

File: addons/account/tests/test_account_tax.py [hunk 6]
```diff
@@ -28,3 +30,98 @@ def test_changing_tax_company(self):
...
+        self.env['account.reconcile.model'].create({
+            'name': "test_tax_is_used",
+            'rule_type': 'writeoff_suggestion',
+            'auto_reconcile': False,
+            'line_ids': [Command.create({
+                'account_id': self.company_data['default_account_revenue'].id,
+                'tax_ids': [Command.set(tax_renconciliation.ids)],
+            })],
+        })
+        tax_renconciliation.invalidate_model(fnames=['is_used'])
```
Comment (qdp-odoo) [inline addons/account/tests/test_account_tax.py:72]: typo ^^

File: addons/account/tests/test_account_tax.py [hunk 7]
```diff
@@ -28,3 +30,98 @@ def test_changing_tax_company(self):
...
+                    'price_unit': 100.0,
+                    'tax_ids': [Command.set(tax_invoice.ids)],
+                }),
+            ],
+        })
+        tax_invoice.invalidate_model(fnames=['is_used'])
+        self.assertTrue(tax_invoice.is_used)
+
+        # Account.reconcile is another of transaction
+        tax_renconciliation = self.env['account.tax'].create({
```
Comment (qdp-odoo) [inline addons/account/tests/test_account_tax.py:59]: typo

**Review by qdp-odoo — CHANGES_REQUESTED** [review]
almost there. I find that approach with the repartition_lines_str much more clean :+1: 

thanks

File: addons/point_of_sale/tests/test_pos_products_with_tax.py [hunk 8]
```diff
@@ -541,3 +541,11 @@ def test_fixed_tax_positive_qty(self):
             {'account_id': self.sale_account.id, 'balance': 0},
             {'account_id': self.cash_pm1.receivable_account_id.id, 'balance': 1},
         ])
+
+    def test_tax_is_used_when_in_transactions(self):
+        ''' Ensures that a tax is set to used when it is part of some transactions '''
+
+        # Call another test that uses product_1
```
[the remark this answers is not in the document]
Comment (aboo-odoo) [reply addons/point_of_sale/tests/test_pos_products_with_tax.py:548]: Indeed, this would truly show that the tax.is_used is correctly updated.

File: addons/account/models/account_tax.py [hunk 9]
```diff
@@ -225,6 +238,84 @@ def _compute_tax_group_id(self):
...
+                    }
+                    repartition_lines_str += str(repartition_line_info) + '//'
+                repartition_lines_str = repartition_lines_str.strip('//')
+            tax.repartition_lines_str = repartition_lines_str
+
+    def _message_log_repartition_lines(self, old_value_str, new_value_str):
+        old_values = old_value_str.split('//')
+        new_values = new_value_str.split('//')
+        if len(old_values) != len(new_values):
+            raise UserError(_("Taxes should have the same amount of invoice and refund repartition lines."))
```
[the remark this answers is not in the document]
Comment (aboo-odoo) [reply addons/account/models/account_tax.py:280]: {old/new}_values is a list in which each item is a repartition line. The error message should in fact be more like: """This tax has already been used, you cannot add nor delete its tax repartition lines."""
This code is only meant to be executed when the tax is used (called from _message_log). I have added a check in case someone else decides (for any reason) to call the function from elsewhere.

[the remark this answers is not in the document]
Comment (qdp-odoo) [reply addons/account/models/account_tax.py:280]: but this error message already exists, in create() and unlink() of tax.repartition.line...

[the remark this answers is not in the document]
Comment (aboo-odoo) [reply addons/account/models/account_tax.py]: What do you think I should be doing if the number of repartition line differs after the modification of a used tax? Should I simply do nothing and return the function?

[the remark this answers is not in the document]
Comment (aboo-odoo) [reply addons/account/models/account_tax.py]: or simply not check at all and assume that the number will always be the same since a tax that is used does not allow for lines to be added and/or removed?

[the remark this answers is not in the document]
Comment (qdp-odoo) [reply addons/account/models/account_tax.py]: > simply not check at all and assume that the number will always be the same since a tax that is used does not allow for lines to be added and/or removed?

yes indeed... I don't see how that could happen, so I don't see the purpose of checking that :)

**Comment by qdp-odoo** [comment]
@aboo-odoo il faut aussi check pq le ci-security est rouge et dire pourquoi tu penses que c'est OK :)

File: addons/account/models/account_tax.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (aboo-odoo) [reply addons/account/models/account_tax.py]: Removed

**Comment by aboo-odoo** [comment]
@odoo/rd-security I'm using a `Markup` from `markupsafe` to display a nicely formatted message in the chatter. The body of the message includes some variables (that could be translated); those have been escaped using the `escape` function from the same library. Thus, I son't see any risk of XSS.

File: addons/account/models/account_tax.py [hunk 10]
```diff
@@ -225,6 +238,93 @@ def _compute_tax_group_id(self):
...
+        new_values = new_value_str.split('//')
+
+        kwargs = {}
+        for old_value, new_value in zip(old_values, new_values):
+            if old_value != new_value:
+                old_value = ast.literal_eval(old_value)
+                new_value = ast.literal_eval(new_value)
+                diff_keys = [key for key in old_value if old_value[key] != new_value[key]]
+                repartition_line = self.env['account.tax.repartition.line'].search([('id', '=', new_value['id'])])
+                body = "<b>{type}</b> {rep} {seq}:<ul class='mb-0 ps-4'>{changes}</ul>".format(
```
Comment (mart-e) [inline addons/account/models/account_tax.py:291]: don't put a `Markup()` around the final result but create a Markup on the actual content, so
```suggestion
                body = Markup("<b>{type}</b> {rep} {seq}:<ul class='mb-0 ps-4'>{changes}</ul>").format(
```
(and you won't need these explicit `escape`

Comment (mart-e) [reply addons/account/models/account_tax.py:291]: @aboo-odoo if `changes` contains html, `Markup` should be put around the markup content. And don't forget that `''.join(Markup)` produces a string, not a Markup. See the doc for some examples:
https://www.odoo.com/documentation/master/developer/reference/backend/security.html#creating-safe-content-using-markup

Comment (aboo-odoo) [reply addons/account/models/account_tax.py]: Suggestion have been applied :smile: Thanks again!

File: addons/account/models/account_tax.py [hunk 11]
```diff
@@ -225,6 +238,93 @@ def _compute_tax_group_id(self):
...
+            if old_value != new_value:
+                old_value = ast.literal_eval(old_value)
+                new_value = ast.literal_eval(new_value)
+                diff_keys = [key for key in old_value if old_value[key] != new_value[key]]
+                repartition_line = self.env['account.tax.repartition.line'].search([('id', '=', new_value['id'])])
+                body = Markup("<b>{type}</b> {rep} {seq}:<ul class='mb-0 ps-4'>{changes}</ul>").format(
+                    type=repartition_line.document_type.capitalize(),
+                    rep=_('repartition line'),
+                    seq=repartition_line.sequence + 1,
+                    changes=Markup('').join(
```
Comment (mart-e) [inline addons/account/models/account_tax.py:295]:
```suggestion
                    changes=Markup().join(
```

File: addons/account/models/account_tax.py [hunk 12]
```diff
@@ -1382,6 +1489,22 @@ class AccountTaxRepartitionLine(models.Model):
...
     tag_ids_domain = fields.Binary(string="tag domain", help="Dynamic domain used for the tag that can be set on tax", compute="_compute_tag_ids_domain")
 
+    @api.model_create_multi
+    def create(self, vals):
+        tax_ids = list(set([line.get('tax_id') for line in vals]))
+        taxes = self.env['account.tax'].search([('id', 'in', tax_ids)])
+        tax_used = dict(taxes.mapped(lambda tax: (tax.id, tax.is_used)))
+        for line in vals:
+            if tax_used.get(line.get('tax_id'), False):
+                raise ValidationError(_("This tax has already been used, you cannot add nor delete its tax repartition lines."))
```
Comment (mart-e) [inline addons/account/models/account_tax.py:1499]: "This tax", which one? You are in a for loop. Also a good idea to say used where if one needs to take action.

**Review by mart-e — COMMENTED** [review]
@robodoo override=ci/security
Reason: Markup for log message

**Comment by qdp-odoo** [comment]
@robodoo r+ rebase-ff

@aboo-odoo thanks, well done!

File: addons/account/models/account_tax.py [hunk 13]
```diff
@@ -225,6 +238,93 @@ def _compute_tax_group_id(self):
...
+            To be overriden to add taxed transactions in the computation of `is_used`
+            Should return a Counter containing a dictionary {record: int} where
+            the record is an account.tax object. The int should be greater than 0
+            if the tax is used in a transaction.
+        '''
+        return Counter()
+
+    def _compute_is_used(self):
+        taxes_in_transactions_ctr = (
+            Counter(dict(self.env['account.move.line']._read_group([], groupby=['tax_ids'], aggregates=['__count']))) +
```
[the remark this answers is not in the document]
Comment (qdp-odoo) [reply addons/account/models/account_tax.py:252]: @xmo-odoo indeed, it looks like we completely overlooked that. Thanks for pointing it out

@aboo-odoo can you make a branch improving that? use sql queries that perform EXISTS, only check for use in other models if not already true. Thanks

[the remark this answers is not in the document]
Comment (aboo-odoo) [reply addons/account/models/account_tax.py:252]: Working on it

[the remark this answers is not in the document]
Comment (aboo-odoo) [reply addons/account/models/account_tax.py:252]: https://github.com/odoo/odoo/pull/139497
